### DISCUSSION TITLE
Feat request: Make the frontend serverless

### DISCUSSION THREAD

**Discussion by johannesschaffer** [payloadcms/payload#1682] [Feature Requests & Ideas]
This way it would me much easier and cheaper to host it on e.g. Vercel & could potentially be deployed alongside/embedded inside the main website. Sanity.io has the option (their admin panel can be serverless deployd) and KeystoneJS is working on it (to embed the admin ui inside a Next.js app). The difficulty and cost related to deploying a node.js server is currently the only thing holding me back from using Payload over Contentful.
File storage would obviously be not possible this way, but cloud adapters to e.g. S3 already exist.

**Comment by mabry1985** [comment]
Yeah, this is why I ended up going with Webiny CMS over Payload. I love how Payload works, but I don't want to pay for a server that is only used for admin tasks. I would 100% switch back to Payload if this was sorted.

**Reply by iakshay** [reply]
Is hosting Webiny cheaper than running Payload on EC2 or DO droplet?

Looking at the [deployment diagram](https://www.webiny.com/docs/architecture/deployment-modes/production), I would be surprised if its cheaper. Are you relying on AWS free limits?

IMO, having a server is cheaper than going serverless for small-ish websites.

**Reply by johannesschaffer** [reply]
Don't know how Webiny is deployed exactly, but if it could be deployed as proposed, to e.g. vercel, like e.g. a react app, it wouldn't cost anything. I'm currently hosting payload at DO App platform for 5€/m. The database is hosted in MongoDB serverless plan for free.

**Reply by iakshay** [reply]
FWIW, I did experiment with deploying payload to Vercel - https://github.com/iakshay/payload-next. 

It mostly works, you just have add redirects to admin route. Other hacks are more for running it locally.

**Reply by johannesschaffer** [reply]
It's extremely slow however, since the vercel functions suspend in between calls, therefore the payload node server has to be started up again, each time. Vercel has an article about deploying node servers and doesn't recommend it

**Reply by mabry1985** [reply]
@iakshay That is a valid point that I would need to research more. I would be leveraging the free tier for my need.

@johannesschaffer it seems to be pretty tied into the AWS ecosystem, but I'm sure with some work you could get it setup as vercel edge functions

**Reply by iakshay** [reply]
Right, I don't think Vercel supports hosting express server on their platform. The only way to host backend seems to be via serverless functions.

It's slow because we are validating config on startup, should be much better with 1.5.0 which skips config validation in production environment.

The NodeJS runtime itself has ~ 250s cold start time, everything else should really be payload. You can play around here - https://lambda-cold-start.vercel.app/

**Reply by johannesschaffer** [reply]
Given that all requests go through the "admin panel" (the server it is hosted on) this startup time is unacceptable (and I think you meant 250ms, right?😅)

**Reply by jmikrut** [reply]
😈 it's almost like 1.5.0 was _leading up to something_

**Reply by iakshay** [reply]
ah, yes 250ms. Will update the demo to see if things get better.

BTW - cold start should only be the first request. It may not be the worst thing if your whole website is statically generated and you only need to hit API for updating content. In those case using Vercel for embedded deployment might be perfect.

In cases where you do need to hit the API's, going with EC2 or DO seems like a better option to me - in terms of both cost and performance.

**Reply by johannesschaffer** [reply]
I wouldn't agree on cost (you get 100GB free per project on Vercel), but yeah, definetly for performance. DO App platform is what I'm doing at the moment. It's not just about the cost, though, but also the DX. Uploading to DO (also tried Heroko) was really a pain (they e.g. have that weird thing to not install dev depencies, which means teh payload scripts can't be run since they rely on e.g. cross-env). Vercel is literally one click, while DO took me 2.5h the first time 🙈

Also, it kinda sucks for development, where you don't pregenerate the sites. Having to wait the cold boot time on every site change completely ruins e.g. page transitions.

Given all that, I stil think it would be a godd idea to make the front-end truly serverless (I mean it's build with Next.js anyways, right? Probably wouldn't be so hard then, to at least give the option for people that don't need the local API?)

**Reply by jamesvclements** [reply]
@johannesschaffer how did you end up solving the cross-env issue for DO? I'm running in to that now trying to deploy on App Platform

**Comment by johannesschaffer** [comment]
I understand, given your local API, that you probably don't wanna make the front end serverless, though 🙈. The only thing I don't quite understand is the benefit of the local API given, that most websites, especially Next.js ones, are deployed serverlessly. I would consider the tradeoffs of making it serverless smaller than having the local API

**Comment by u11d-michal-miler** [comment]
This discussion really highlights the core tension with Payload and serverless platforms.

Even if you manage to deploy it on Vercel/Amplify/etc., the real issues usually show up later: cold starts, multiple concurrent initializations, DB connection pressure, and general instability around long-lived CMS behavior. Payload still fundamentally behaves like an Express app that expects a persistent runtime, not ephemeral functions.

We ran into the same class of problems and ended up moving away from serverless for the CMS layer entirely. I documented a working approach using AWS Amplify + MongoDB Atlas that avoids most of these constraints while still keeping infrastructure relatively simple and free-tier friendly:

https://u11d.com/blog/how-to-deploy-payload-cms-on-aws-amplify-with-atlas-mongo-for-free/

Might help anyone trying to stay on serverless but running into instability rather than outright build errors.
